internal/backup/csi_volumesnapshot.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2020 the Velero contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package backup
+
+import (
+	snapshotv1beta1api "github.com/kubernetes-csi/external-snapshotter/v2/pkg/apis/volumesnapshot/v1beta1"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/vmware-tanzu/velero-plugin-for-csi/internal/util"
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	"github.com/vmware-tanzu/velero/pkg/plugin/velero"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// VolumeSnapshotBackupItemAction is a backup item action plugin to backup
+// CSI VolumeSnapshot objects using Velero
+type VolumeSnapshotBackupItemAction struct {
+	Log logrus.FieldLogger
+}
+
+// AppliesTo returns information indicating that the CSISnapshotter action should be invoked to backup volumesnapshots.
+func (p *VolumeSnapshotBackupItemAction) AppliesTo() (velero.ResourceSelector, error) {
+	p.Log.Info("VolumeSnapshotBackupItemAction AppliesTo")
+
+	return velero.ResourceSelector{
+		IncludedResources: []string{"volumesnapshots.snapshot.storage.k8s.io"},
+	}, nil
+}
+
+func setVolumeSnapshotAnnotationsAndLabels(vs *snapshotv1beta1api.VolumeSnapshot, vals map[string]string) {
```

Comment:
not a fan of duplicating the info in annotations and labels.

I would prefer to put all the metadata we generate into annotations, unless users may want to use the labels to query. WDYT?

Reply:
In general I've found labels to be more useful for expressing relationships, since you can query on them - for example, we use labels to relate PodVolumeBackups to Backups, which makes it easy to query all PodVolumeBackups that are related to a Backup for CLI display purposes, etc.

Other than that, if it's not something we're going to filter on, then I'd say an annotation make sense.

Reply:
I'd echo what Steve said - if it's something we might query/filter on, a label makes sense. Otherwise, it can be on an annotation.

I'm not sure if there's different storage limitations between the two or not.